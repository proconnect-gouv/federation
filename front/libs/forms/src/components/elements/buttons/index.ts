export * from './array-add';
export * from './array-remove';
export * from './clipboard';
