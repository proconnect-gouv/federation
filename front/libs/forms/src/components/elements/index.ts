export * from './fieldset';
export * from './fieldset-legend';
export * from './group';
export * from './label';
export * from './message';
