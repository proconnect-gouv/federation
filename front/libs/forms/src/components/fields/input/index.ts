/* istanbul ignore file */

// declarative file
export * from './input.field';
