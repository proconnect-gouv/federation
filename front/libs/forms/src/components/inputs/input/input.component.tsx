import React from 'react';
import type { FieldInputProps } from 'react-final-form';

import type { PropsWithClassName } from '@fc/common';

interface InputComponentProps<FieldValue = string> extends PropsWithClassName {
  readOnly?: boolean;
  disabled?: boolean;
  id: string;
  input: FieldInputProps<FieldValue, HTMLElement | HTMLSelectElement>;
}

export const InputComponent = React.memo(
  ({ className, disabled = false, id, input, readOnly = false }: InputComponentProps) => (
    <input
      {...input}
      aria-describedby={`${input.name}-messages`}
      className={className}
      data-testid={`${id}--testid`}
      disabled={disabled}
      id={id}
      name={input.name}
      readOnly={readOnly}
      type={input.type}
    />
  ),
);

InputComponent.displayName = 'InputComponent';
