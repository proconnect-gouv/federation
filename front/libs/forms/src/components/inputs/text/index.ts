/* istanbul ignore file */

// declarative file
export * from './text.input';
