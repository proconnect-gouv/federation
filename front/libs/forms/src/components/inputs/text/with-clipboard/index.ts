export * from './with-clipboard.input';
