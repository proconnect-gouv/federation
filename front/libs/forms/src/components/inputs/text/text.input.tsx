import React from 'react';

import { ComponentTypes } from '../../../enums';
import { useFieldMeta } from '../../../hooks';
import type { PropsWithInputConfigType } from '../../../types';
import { GroupElement, LabelElement, MessageElement } from '../../elements';

export const TextInput = React.memo(({ config, input, meta }: PropsWithInputConfigType) => {
  const { hint, label, required } = config;
  const { className, disabled, name, type } = input;

  const { errorMessage, hasError, inputClassname, isValid } = useFieldMeta(meta);

  const id = `form-input-text-${name}`;
  return (
    <GroupElement
      className={className}
      disabled={disabled}
      hasError={hasError}
      isValid={isValid}
      type={ComponentTypes.INPUT}>
      <LabelElement hint={hint} label={label} name={name} required={required} />
      <input
        {...input}
        aria-describedby={`${name}-messages`}
        className={inputClassname}
        data-testid={`${id}--testid`}
        disabled={disabled}
        id={id}
        name={name}
        type={type}
      />
      <MessageElement error={errorMessage} id={name} isValid={isValid} />
    </GroupElement>
  );
});

TextInput.displayName = 'TextInput';
