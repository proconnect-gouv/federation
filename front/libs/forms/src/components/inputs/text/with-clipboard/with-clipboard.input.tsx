import classnames from 'classnames';
import React, { useCallback } from 'react';
import type { FieldInputProps } from 'react-final-form';

import { type PropsWithClassName, useClipboard } from '@fc/common';

import { ClipboardButton } from '../../../elements';
import { InputComponent } from '../../input';
import classes from './with-clipboard.module.scss';

interface InputWithClipboardProps extends PropsWithClassName {
  id: string;
  input: FieldInputProps<string, HTMLElement>;
  inputClassname?: string | undefined;
}

export const InputWithClipboard = React.memo(
  ({ className, id, input, inputClassname }: InputWithClipboardProps) => {
    const { value } = input;
    const { onCopy } = useClipboard();

    const copyHandler = useCallback(() => {
      onCopy(value);
    }, [value, onCopy]);

    return (
      <div className={classnames('flex-columns', className)}>
        <InputComponent
          readOnly
          className={classnames(inputClassname, classes.withClipboard)}
          id={id}
          input={input}
        />
        <ClipboardButton dataTestId={`${id}-copy-button`} onClick={copyHandler} />
      </div>
    );
  },
);

InputWithClipboard.displayName = 'InputWithClipboard';
