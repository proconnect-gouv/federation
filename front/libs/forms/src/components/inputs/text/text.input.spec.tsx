import { render } from '@testing-library/react';

import { useFieldMeta } from '../../../hooks';
import type { PropsWithInputConfigType } from '../../../types';
import { GroupElement, LabelElement, MessageElement } from '../../elements';
import { InputComponent } from '../input';
import { TextInput } from './text.input';
import { InputWithClipboard } from './with-clipboard';

jest.mock('./with-clipboard/with-clipboard.input');
jest.mock('../input/input.component');
jest.mock('../../elements/group/group.element');
jest.mock('../../elements/label/label.element');
jest.mock('../../elements/message/message.element');
jest.mock('../../../hooks/field-meta/field-meta.hook');

describe('TextInput', () => {
  // Given
  const isValidMock = Symbol('isValidMock') as unknown as boolean;
  const hasErrorMock = Symbol('hasErrorMock') as unknown as boolean;
  const touchedMock = Symbol('touchedMock') as unknown as boolean;
  const readonlyMock = Symbol('readonlyMock') as unknown as boolean;

  const propsMock = {
    config: {
      hint: 'any-hint-mock',
      label: 'any-label-mock',
      readonly: readonlyMock,
    },
    input: {
      className: 'any-classname-mock',
      name: 'any-name-mock',
      type: 'text',
      value: 'any-input-value-mock',
    },
    meta: {
      error: 'any-errorMessage-mock',
      touched: touchedMock,
      value: 'any-input-value-mock',
    },
  } as unknown as PropsWithInputConfigType;

  beforeEach(() => {
    // Given
    jest.mocked(useFieldMeta).mockReturnValueOnce({
      errorMessage: 'any-errorMessage-mock',
      hasError: hasErrorMock,
      inputClassname: 'any-inputClassname-mock',
      isValid: isValidMock,
    });
  });

  it('should match snapshot', () => {
    // When
    const { container } = render(<TextInput {...propsMock} />);

    // Then
    expect(container).toMatchSnapshot();
    expect(useFieldMeta).toHaveBeenCalledOnce();
    expect(useFieldMeta).toHaveBeenCalledWith({
      error: 'any-errorMessage-mock',
      touched: touchedMock,
      value: 'any-input-value-mock',
    });
    expect(GroupElement).toHaveBeenCalledOnce();
    expect(GroupElement).toHaveBeenCalledWith(
      {
        children: expect.any(Array),
        className: 'any-classname-mock',
        hasError: hasErrorMock,
        isValid: isValidMock,
        type: 'input',
      },
      {},
    );
    expect(LabelElement).toHaveBeenCalledOnce();
    expect(LabelElement).toHaveBeenCalledWith(
      {
        hint: 'any-hint-mock',
        label: 'any-label-mock',
        name: 'any-name-mock',
      },
      {},
    );
    expect(MessageElement).toHaveBeenCalledOnce();
    expect(MessageElement).toHaveBeenCalledWith(
      {
        dataTestId: 'any-name-mock-messages',
        error: 'any-errorMessage-mock',
        id: 'any-name-mock',
        isValid: isValidMock,
      },
      {},
    );
  });

  it('should match snapshot, when readonly is true', () => {
    // Given
    const propsReadOnlyMock = {
      ...propsMock,
      config: {
        hint: 'any-hint-mock',
        label: 'any-label-mock',
        readonly: true,
      },
    };

    // When
    const { container } = render(<TextInput {...propsReadOnlyMock} />);

    // Then
    expect(container).toMatchSnapshot();
    expect(InputWithClipboard).toHaveBeenCalledOnce();
    expect(InputWithClipboard).toHaveBeenCalledWith(
      {
        className: 'fr-mt-1w',
        id: 'form-input-text-any-name-mock',
        input: propsMock.input,
        inputClassname: 'any-inputClassname-mock',
      },
      {},
    );
  });

  it('should match snapshot, when readonly is false', () => {
    // Given
    const propsReadOnlyMock = {
      ...propsMock,
      config: {
        hint: 'any-hint-mock',
        label: 'any-label-mock',
        readonly: false,
      },
    };

    // When
    const { container } = render(<TextInput {...propsReadOnlyMock} />);

    // Then
    expect(container).toMatchSnapshot();
    expect(InputComponent).toHaveBeenCalledOnce();
    expect(InputComponent).toHaveBeenCalledWith(
      {
        className: 'any-inputClassname-mock',
        id: 'form-input-text-any-name-mock',
        input: propsMock.input,
      },
      {},
    );
  });
});
