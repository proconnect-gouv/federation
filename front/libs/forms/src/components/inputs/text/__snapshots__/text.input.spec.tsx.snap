// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput should match snapshot 1`] = `
<div>
  <div
    data-mockid="GroupElement"
  >
    <div
      data-mockid="LabelElement"
    />
    <div
      data-mockid="InputWithClipboard"
    />
    <div
      data-mockid="MessageElement"
    />
  </div>
</div>
`;

exports[`TextInput should match snapshot, when readonly is false 1`] = `
<div>
  <div
    data-mockid="GroupElement"
  >
    <div
      data-mockid="LabelElement"
    />
    <div
      data-mockid="InputComponent"
    />
    <div
      data-mockid="MessageElement"
    />
  </div>
</div>
`;

exports[`TextInput should match snapshot, when readonly is true 1`] = `
<div>
  <div
    data-mockid="GroupElement"
  >
    <div
      data-mockid="LabelElement"
    />
    <div
      data-mockid="InputWithClipboard"
    />
    <div
      data-mockid="MessageElement"
    />
  </div>
</div>
`;
